File: app/api/auth/auth.service.go
package auth

import (
	"encoding/json"
	"errors"

	"github.com/gDenisLit/item-server-go/cmd/api/user"
	"github.com/gDenisLit/item-server-go/cmd/dtos"
	"github.com/gDenisLit/item-server-go/cmd/models"
	"github.com/gDenisLit/item-server-go/cmd/services"
	"golang.org/x/crypto/bcrypt"
)

func Login(credentials *dtos.LoginDTO) (*models.User, error) {
	user, err := user.GetByUsername(credentials.Username)
	if err != nil {
		return nil, err
	}

	err = bcrypt.CompareHashAndPassword(
		[]byte(user.Password),
		[]byte(credentials.Password),
	)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func Signup(credentials *dtos.SignupDTO) (*models.User, error) {
	_, err := user.GetByUsername(credentials.Username)
	if err == nil {
		return nil, errors.New("username already taken")
	}

	saltRounds := 10
	hash, err := bcrypt.GenerateFromPassword(
		[]byte(credentials.Password),
		saltRounds,
	)
	if err != nil {
		return nil, err
	}

	credentials.Password = string(hash)
	user, err := user.Add(credentials)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func GetLoginToken(user *dtos.UserDTO) (string, error) {

	userJson, err := json.Marshal(user)
	if err != nil {
		return "", err
	}

	token, err := services.Encode("loginToken", userJson)
	if err != nil {
		return "", err
	}
	return token, nil
}

func ValidateToken(token string) (*models.User, error) {

	return nil, nil
}

File: app/api/auth/auth_service.py
import bcrypt
import jwt
from jwt.exceptions import DecodeError
from app.api.user.user_service import UserService
from app.models.User_model import User
from app.config import config
from typing import Optional


class AuthService:

    @staticmethod
    async def login(credentials: dict[str, str]) -> User:
        error_msg = "Wrong credentials"

        user = await UserService.get_by_username(credentials["username"])
        if not user:
            raise ValueError(error_msg)

        correct = bcrypt.checkpw(
            credentials["password"].encode('utf-8'),
            user.password.encode('utf-8')
        )

        if not correct:
            raise ValueError(error_msg)

        return user

    @staticmethod
    async def signup(credentials: dict[str, str]) -> User:

        user = await UserService.get_by_username(credentials["username"])
        if user:
            raise ValueError("username is already taken")

        hashed_password = bcrypt.hashpw(
            credentials["password"].encode(),
            bcrypt.gensalt(10)
        )
        credentials["password"] = hashed_password
        return await UserService.add_user(credentials)

    @staticmethod
    async def get_login_token(user: User) -> str:
        token = jwt.encode(
            user.to_dict(),
            config.secret_key,
            algorithm="HS256"
        )
        return token

    @staticmethod
    async def validate_token(token: str) -> Optional[dict]:
        try:
            user = jwt.decode(
                token,
                config.secret_key,
                algorithms=["HS256"]
            )
            return user
        except DecodeError:
            raise ValueError("Invalid login token")
